src/tmpfiles/tmpfiles.c

```diff
@@ -857,13 +864,20 @@ static int path_set_perms(Item *i, const char *path) {
                 return r;
         }
 
-        if (fstat(fd, &st) < 0)
-                return log_error_errno(errno, "Failed to fstat() file %s: %m", path);
+        if (i->type == EMPTY_DIRECTORY) {
+                struct stat stbuf;
+
+                /* FIXME: introduce fd_is_dir() helper ? */
+                if (fstat(fd, &stbuf) < 0)
+                        return log_error_errno(errno, "Failed to fstat() file %s: %m", path);
 
-        if (i->type == EMPTY_DIRECTORY && !S_ISDIR(st.st_mode))
-                return log_error_errno(EEXIST, "'%s' already exists and is not a directory. ", path);
+                if (!S_ISDIR(stbuf.st_mode))
+                        return log_error_errno(EEXIST, "'%s' already exists and is not a directory. ", path);
```

Comment:
so, i see this is not introduced by this patch, but so far we had the rule that we the error codes we pass to log_error_errno() and friends is only the error codes that indicate errors we *received* from somewhere else, not the ones we *generate* ourselves and propagate elsewhere. Hence I figure this should be rewritten to:

```
if (!S_ISDIR(stbuf.st_mode)) {
        log_error("'%s' already exists … 
        return -EEXIST;
}
```